servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/RetryingHttpRequesterFilter.java

```diff
@@ -258,19 +295,31 @@ protected Single<StreamingHttpResponse> request(final StreamingHttpRequester del
             if (responseMapper != null) {
                 single = single.flatMap(resp -> {
                     final HttpResponseException exception = responseMapper.apply(resp);
-                    return (exception != null ?
-                            // Drain response payload body before discarding it:
-                            resp.payloadBody().ignoreElements().onErrorComplete()
-                                    .concat(Single.<StreamingHttpResponse>failed(exception)) :
-                            Single.succeeded(resp))
-                            .shareContextOnSubscribe();
+                    if (exception == null) {
+                        return Single.succeeded(resp).shareContextOnSubscribe();
+                    }
+                    if (returnFailedResponses) {
+                        return Single.failed(new WrappedResponseException(resp, exception));
```

Comment:
We don't need another wrapper over the existing `HttpResponseException`. Consider unwrapping `HttpResponseException` inside the final `onErrorResume`.